### PR TITLE
{Core} Query examples: add a configurable global parameter that can recommend JMESPath query strings.

### PR DESCRIPTION
**Pull request by Edmund-Shi**
## Description

Most `show` or `list` commands return a JSON output which may be quite long and difficult to understand. So `az` provide the `--query` argument to filter the output with JMESPath string. But JMESPath language is different from any other programming language. So writing a JMESPath string is a hard job for beginners.

The main purpose of this project is to automatically generate JMESPath query string based on commands output. We add a global argument `--query-examples` for `az` command. When you add this argument to any command, we will run this command, and then analyze command output. Finally, give you a list of JMESPath query string that you can use. Then you just need to copy the recommendations and paste them after `--query` arguments to get the filtered results. Also, you can customize your query string based on the examples.

### What's new in this PR

This is the second and the final part of the feature of `query-examples`. `az` command will read config file, and add `query-examples` parameter to commands that in the allow-list.  Commands that end with `list` and `show` are in the allow-list by default. You can turn off this feature by setting allow-list to empty using `az config set query.allow_list=""`.

Other features:

* Recommend query examples based on your positional arguments.
* Recommend commonly used JMESPath syntax.
* Config the number of examples via `az config` or the config file.

### Unit Tests:

You can find unit test in `src/azure-cli-core/azure/cli/core/tests/test_query_examples.py`. We provide some sample JSON, and check whether the tree is constructed as expected. We also check whether correct JEMSPath strings are returned.

Here are the unit tests list:
* Tests for `QueryTreeBuilder`

Test   Inputs | Expected   Results | Example   Input
-- | -- | --
A simple JSON dict | `builder._root` is not `None` <br/>`builder._all_nodes` is not None  <br/>`builder._all_nodes` is instance of dict <br/>Corresponding JSON keys are in `builder._all_nodes` |  image
A simple JSON list | `builder._root` is not `None`<br/> `builder._all_nodes` is not `None`<br/>`builder._all_nodes` is instance of dict<br/>Corresponding JSON keys are in `builder._all_nodes`<br/>`builder._root._is_array` is True |  image
An empty dict | `builder._root` is not `None`.<br/>`builder._root._child` is empty. | `{}`
An empty list | `builder._root` is None. (Because there is no node found.) | `[]`
A list with empty dict | `builder._root` is not `None`. <br/>`builder._root._child` is empty. | `[{},{}]`

* Tests for `generate_examples`

Test   Inputs | Expected   Results | Example   Input
-- | -- | --
A simple JSON dict | Select query string for all fields. | image
A simple JSON list | Select query string and filter query string and query string with some functions. |  image
A simple dict   A list of keywords | Only return query strings that match the input keywords | 
Change the default configs | Change `max_examples`, limit the number of outputs. | A big JSON.

* Test for `QueryExample`:

Test   Inputs | Expected   Results | Example   Input | Example Results
-- | -- | -- | -- 
A query string with backticks in it. | Escape this backtick. |  ``[?contains(name,'Ann')==`true`]`` | ``[?contains(name,'Ann')==\`true\`]``
Limit the length of query string or help string | The max length of the string is limited. | `This.is.a.very.very.very.long.query.string` | `This.is.a.very.ve...`

### Scenario Tests:

Expected outputs are examples of JMESPath query strings with help messages. The query strings depends on the outputs of the command which may differ for different users.

- Test with default configs:

```bash
$ az group list --query-examples
Query string                                               Help
---------------------------------------------------------  -----------------------------------------------
[].type                                                    Show the value of type field.
[?type=='Microsoft.Resources/resourceGroups']              Show the resources that satisfy the condition.
[?contains(@.type, 'something')==\`true\`].type            Show the type field that contains given string.
[].tags.type                                               Show the value of type field.
[?tags.type=='test']                                       Show the resources that satisfy the condition.
[?contains(@.tags.type, 'something')==\`true\`].tags.type  Show the type field that contains given string.
[].name                                                    Show the value of name field.
[?name=='xxxxxxxxxx-rg-MyWorkspace-xxxxxxxxxxxxx']         Show the resources that satisfy the condition.
[?contains(@.name, 'something')==\`true\`].name            Show the name field that contains given string.
[].location                                                Show the value of location field.

# get examples based on custom keywords
$ az vm list --query-examples location
Query string                                             Help
-------------------------------------------------------  ---------------------------------------------------
[].location                                              Show the value of location field.
[?location=='eastus']                                    Show the resources that satisfy the condition.
[?contains(@.location, 'something')==\`true\`].location  Show the location field that contains given string.
```

- Test `allow_list`:

```bash
# disable --query-examples for all commands
$ az config set query.allow_list=""
$ az vm list --query-examples
az: error: unrecognized arguments: --query-examples

# enable --query-examples for custom commands
$ az config set query.allow_list="list,show,version"
$ az version --query-examples
Query string                     Help
-------------------------------  ---------------------------------------------
azure-cli-core                   Show the value of azure-cli-core field.
azure-cli                        Show the value of azure-cli field.
extensions.kusto                 Show the value of kusto field.
extensions                       Show the value of extensions field.

# change back to default
$ az config unset query.allow_list
```

- Test output format:

```bash
# set max length for help string and query string
$ az config set query.help_len=20
$ az config set query.examples_len=120
# limit the number of output, set to -1 to print all examples
$ az config set query.max_examples=5
$ az vm list --query-examples
Query string                                                                                                       Help
-----------------------------------------------------------------------------------------------------------------  -----------------------
[].storageProfile.imageReference.version                                                                           Show the value of ve...
[?storageProfile.imageReference.version=='latest']                                                                 Show the resources t...
[?contains(@.storageProfile.imageReference.version, 'something')==\`true\`].storageProfile.imageReference.version  Show the version fie...
[].storageProfile.imageReference.offer                                                                             Show the value of of...
[?storageProfile.imageReference.offer=='UbuntuServer']                                                             Show the resources t...

# print output in another format
$ az vm list --query-examples -o tsv
Show the value of vmId field.   [].vmId
Show the resources that satisfy the condition.  [?vmId=='46f8c9f5-xxxx-xxxx-xxxx-xxxxxxxxxxxx']
Show the vmId field that contains given string. [?contains(@.vmId, 'something')==\`true\`].vmId
Show the value of resourceGroup field.  [].storageProfile.osDisk.managedDisk.resourceGroup
Show the resources that satisfy the condition.  [?storageProfile.osDisk.managedDisk.resourceGroup=='BIM_XXXXXXX_RG']
```

### Architecture Overview

Here is the architecture diagram. We register some event-handler in Azure-CLI Core.
Architecture Overview

- `EVENT_INVOKER_PRE_LOAD_ARGUMENTS`
    - For this event, we register `--query-examples` as an experimental argument for commands that in the allow-list in command table in command loader.
- `EVENT_INVOKER_POST_PARSE_ARGS`
    - For this event, we check if `--query-examples` is in user's input. And register a handler for generating recommendation.
- `EVENT_INVOKER_FILTER_RESULT`
    - For this event, we analyze the output JSON file. And generate recommends as new outputs.

### Query Tree Node
We analyze the output JSON file using a Tree Parser. The parser analyzes the JSON and then build a tree. Each tree node refer to a JSON object or an array. Here is the UML schema for `QueryTreeNode`.

QueryTreeNode

Here is the meaning of each field and method:

<details>
<summary>Fields</summary>

 - `name` is the name of current JSON object. The name of the outermost object is 'root'.
 - `parent` is also a TreeNode which is the parent of current node.
 - `child` is a list of TreeNode which are the children of current node.
 - `data` is a list of raw JSON data. Each item in this list is the JSON object in the same level with same name.
 - If `under_array` is true, then this TreeNode is a converted from a list.
 - `is_dummy` is `True` only in root node.

</details>

<details>
<summary>Methods</summary>

  - `update_node_data` updates node with input data.
 - `add_child` appends `child_node` to `self.child`
 - `get_trace_to_array` gets trace to the nearest array. Return None if no array found. This function return two values, inner trace and outer trace. The inner trace is the trace inside the brackets and the outer trace is the trace outside the brackets.
 - `get_trace_to_root` returns the trace string(aka, select string in JMESPath) to the root node.
 - `get_examples` is the entry function for recommend examples for current node.
 - `get_help_str` returns a help string.
 - `get_filter_str` returns the kind of JMESPath that filters the results with some conditions.
 - `get_contains_str` returns the JMESPath that filter the results with a contains function.
</details>

Let's convert following JSON to TreeNode structure for example.  

```json
{
    "Name": "John",
    "Phone": [
        {
            "type": "home",
            "number": "212 555-1234"
        },
        {
            "type": "office",
            "number": "646 555-4567"
        }
    ]
}
```

The whole JSON will be parsed into a root node as following:

```python
{
    is_dummy: True,
    _data: None,
    _name: '',
    under_array: False,
    _parent: None,
    _child:[
        {
            is_dummy: False,
            _data: ['John'],
            _name: 'Name',
            _under_array: False,
            _parent: <Node ''>,
            _child: []
        },
        {
            is_dummy: False,
            _data: None,
            _name: 'Phone',
            _under_array: True,
            _parent: <Node ''>,
            _child: [
                {
                    is_dummy: False,
                    _data: ['home', 'office'],
                    _name: 'type',
                    _under_array: False,
                    _parent: <Node 'Phone'>,
                    _child: []
                },
                {
                    is_dummy: False,
                    _data: ['212 555-1234', '646 555-4567'],
                    _name: 'number',
                    _under_array: False,
                    _parent: <Node 'Phone'>,
                    _child: []
                }
            ]
        }
    ]
}
```

**Data Storage**

- To avoid data redundancy, we only store JSON data in leaf nodes. And the values in `_data` field are always basic data type such as string and int. In this case, `node 'Phone'` has two child nodes and the value of `Phone` is a list of dict. So the `_data` field for `node 'Phone'` is `None`. Similarly, the `_data` for root node is also `None`.
- A leaf node stores all data in the same level with the same name. In the example above, `node 'type'` stores both `"home"` and `"office"`.

**Trace To Root**

Trace to root describe the JMESPath that can access to current node. For `node 'Name'`, the trace to root is `Name`. For `node 'number'`, the trace to root is `Phone[].number`. 

**Trace To Array**

We divide trace to array into two parts, inner trace and outer trace. The inner trace is the trace to the nearest ancestor that is `_under_array`. The outer trace is the trace to root for the nearest array node. For example, the outer trace and inner trace for `node 'number'`is `Phone` and `number`.  For `node 'Name'`, we can't find any ancestor that is `_under_array`. So both outer trace and inner trace are `None` for `node 'Name'`.

### Query Tree Builder

We parse JSON into TreeNodes within a tree parser. In this parser, we start from JSON root object. And parse all of its child object recursively. Let's have a look at the UML for a `QueryTreeBuilder` :

Query Tree Builder

Here is the meaning of each field and method:

<details>
<summary>Fields:</summary>

 - `_root` : This the root node of the tree. It must be a dummy node. `_root` node can be either `under_array` or not `under_array`.
 - `_all_nodes` : The key of this dict is a string that is corresponding to the key in the JSON. The value of this dict is a list of `TreeNode` which have the same `name`.
 - `_config` : The `CLIConfig`.
</details>
<details>
<summary>Methods</summary>

- `build`: Build a query tree with input JSON. The data should be an object that loads from a JSON.
- `generate_examples` : Generate JMESPath query string examples based on `keyword_list`. If no keyword is provided by user, the `keyword_list` is an empty list. `output_format` is a string that indicate the output format, such as 'table', 'json', etc.
- `update_config` : Set or update current config. Called in `__init__()`.
- `_get_match_node` : Return a list of keys that match the user provided keywords. Return all available keys if no keyword is provided.
- `_parse` : Parse a single node. The `data` argument is a list of data in the JSON with the same depth and the same name. It must be a list. The `under_array` is `True` if the value of this JSON node is an array.
- `_parse_list`: Parse a node whose value is an array.
- `_parse_dict` : Parse a dict.
- `_parse_leaf` : Parse a leaf node that has no child.
- `_get_all_keys` : Get all keys in a list of dict.
- `_get_not_none_value` : Get all not None values in a list of dict for a given key. This function perform `get` on each dict in the list, but only returns not `None` values.
- `_record_node` : Add name and corresponding node to `self._all_nodes`.
</details>

Let's take the previous JSON for illustration. The expected tree builder after call `build` is :

```python
{
    _all_nodes: {
        '': <Node '', aka, the root node>,
        'type': <Node 'type'>,
        'number': <Node 'number'>,
        'Phone': <Node 'Phone'>,
        'Name': <Node 'Name'>
    }
    _config: {'examples_len': 80, 'help_len': 80, 'max_examples': 10},
    _root: <Node ''>
}
```

**History Notes**  

{Core} Add allow-list to control the scope of `query-example`.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
query-examples

File: src/azure-cli-core/azure/cli/core/commands/query_examples.py [hunk 1]
```diff
@@ -56,19 +67,237 @@ def register_query_examples(cli, **kwargs):
         events.EVENT_INVOKER_POST_PARSE_ARGS, handle_example_parameter)
 
 
-class TreeBuilder:
-    '''Parse entry. Generate parse tree from json. And then give examples.
-    '''
+class Example:
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/commands/query_examples.py:70]: QueryExample

File: src/azure-cli-core/azure/cli/core/commands/query_examples.py [hunk 2]
```diff
@@ -56,19 +67,237 @@ def register_query_examples(cli, **kwargs):
...
+            return self._all_nodes.keys()
+        for pattern in keywords_list:
+            for node_name in self._all_nodes:
+                if node_name not in match_list and name_match(pattern, node_name):
+                    match_list.append(node_name)
+
+        return match_list
+
+    def _do_parse(self, name, data, parent, under_array=False):
+        '''do parse for a single node
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/commands/query_examples.py:239]: I think the word `do` is redundant.

**Comment by jiasli** [comment]
Please use more meaningful title instead of "{Core} Finish the development of query-examples".

Also, let's add some manual **test guides** with input and output.

File: src/azure-cli-core/azure/cli/core/commands/query_examples.py [hunk 3]
```diff
@@ -56,19 +70,256 @@ def register_query_examples(cli, **kwargs):
...
+            if value:
+                return value
+        return None
+
+    def get_help_str(self, help_type):
+        """Return help string based on help_type."""
+        help_table = {
+            'contains': "Display {} field that contains given string.".format(self._name),
+            'filter': "Display resources that satisfy the condition.",
+            'select': "Display value of {} field".format(self._name),
```
Comment (Edmund-Shi) [inline src/azure-cli-core/azure/cli/core/commands/query_examples.py:127]: missing `the`

**Review by jiasli — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]

**Review by jsntcy — APPROVED** [review]

**Pull request merged**